### PR TITLE
feat: 🎸 filter sessions by target

### PR DESCRIPTION
**Pull request by lisbet-alvarez**
:tickets: [Jira ticket](https://hashicorp.atlassian.net/browse/ICU-12066)

## Description

Add target filtering on sessions list view page.

## Screenshots (if appropriate):

https://github.com/hashicorp/boundary-ui/assets/107949262/8a800dee-9276-483e-b507-c805424ecf0a

## How to test

1. Use llb-pagination-2 be branch and start boundary dev
2. Copy newly made binary into ui/desktop/electron-app/cli
3. Checkout this branch and run ENABLE_MIRAGE=false BYPASS_CLI_SETUP=true yarn start:desktop

## Checklist:

- [X] I have added before and after screenshots for UI changes
- ~[ ] I have added JSON response output for API changes~
- ~[ ] I have added steps to reproduce and test for bug fixes in the description~
- [X] I have commented on my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- ~[ ] I have added tests that prove my fix is effective or that my feature works~

### CODE REVIEW

File: ui/desktop/app/templates/scopes/scope/projects/sessions/index.hbs [hunk 1]
```diff
@@ -13,110 +13,52 @@
...
-                @name={{filter.name}}
-                @items={{filter.allowedValues}}
-                @selectedItems={{filter.selectedValue}}
-                @onChange={{route-action 'filterBy' filter.name}}
-                as |group|
+  <page.body class='search-filtering'>
+    {{#if this.sortedSessions}}
+      <Hds::SegmentedGroup as |S|>
+        <S.Generic>
+          <Dropdown
```
Comment (ZedLi) [inline ui/desktop/app/templates/scopes/scope/projects/sessions/index.hbs:20]: I don't think I see anything that would necessitate building your own options, I think this could just be:
```
          <Dropdown
            @name={{t 'resources.target.title'}}
            @itemOptions={{this.availableTargets}}
            @checkedItems={{this.targets}}
            @applyFilter={{fn this.applyFilter 'targets'}}
            @isSearchable={{true}}
            @listPosition='bottom-left'
          />
```

Since the component handles building the items for you if you provide the item options

File: ui/desktop/app/routes/scopes/scope/projects/sessions/index.js [hunk 2]
```diff
@@ -39,53 +34,43 @@ export default class ScopesScopeProjectsSessionsIndexRoute extends Route {
...
+    const queryOptions = { query: { filters }, force_refresh: true };
+    const sessions = await this.store.query('session', queryOptions);
 
-    // If status filters are selected, filter sessions by the selected statuses
-    if (this.status?.length) {
-      sessions = sessions.filter((s) => this.status.includes(s?.status));
+    // Query all sessions and all targets for defining filtering values if entering route for the first time
+    if (from !== 'scopes.scope.projects.sessions.index') {
+      const options = { pushToStore: false };
+      this.allSessions = await this.store.query(
```
Comment (ZedLi) [inline ui/desktop/app/routes/scopes/scope/projects/sessions/index.js:56]: What is `allSessions` used for?

Comment (lisbet-alvarez) [reply ui/desktop/app/routes/scopes/scope/projects/sessions/index.js:56]: it is used in the controller file index.js line 37 to make sure that all target filter options are available for existing sessions.

File: ui/desktop/app/templates/scopes/scope/projects/sessions/index.hbs (hunk 1 above)
Comment (lisbet-alvarez) [reply ui/desktop/app/templates/scopes/scope/projects/sessions/index.hbs:20]: fixed, thank u!

Comment (ZedLi) [reply ui/desktop/app/templates/scopes/scope/projects/sessions/index.hbs:20]: I think you'll need to update the component though since it assumes `displayName` and change it in a few places, we should just manually do the `name || id` display rather than assume it's a model with a `displayName`

File: addons/core/addon/components/dropdown/index.hbs [hunk 3]
```diff
@@ -35,7 +40,11 @@
         checked={{includes itemOption.id @checkedItems}}
         {{on 'change' this.selectItem}}
       >
-        {{itemOption.displayName}}
+        {{#if itemOption.name}}
+          {{itemOption.name}}
+        {{else}}
+          {{itemOption.id}}
+        {{/if}}
```
Comment (ZedLi) [inline addons/core/addon/components/dropdown/index.hbs:47]:
```suggestion
        {{#if itemOption.name itemOption.name itemOption.id}}
```

You also need to update this in the targets controller where we make objects with `displayName`

File: ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js [hunk 4]
```diff
@@ -4,25 +4,59 @@
...
+   * with a limit of 100 targets rendered.
+   * @returns {[TargetModel]}
+   */
+  get availableTargets() {
+    const uniqueSessionTargetIds = new Set(
+      this.model.allSessions.map((session) => session.target_id),
+    );
+    return this.model.allTargets
+      .filter((target) => uniqueSessionTargetIds.has(target.id))
+      .slice(0, 100);
```
Comment (ZedLi) [inline ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:42]: I think you should do this in a separate PR, we should be doing this in the drop down component but this will also prevent you from searching all possible targets right?

Comment (lisbet-alvarez) [reply ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:42]: Ah yup you're correct, and good point I'll remove this

File: ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js [hunk 5]
```diff
@@ -4,25 +4,58 @@
...
-  get sorted() {
-    const sessions = this.model;
-    // Sort sessions by created time descending (newest on top)
-    const sortedSessions = orderBy(sessions, 'created_time', 'desc');
-    return [
-      // then move active sessions to the top
-      ...sortedSessions.filter((session) => session.isAvailable),
-      // and all others to the end
-      ...sortedSessions.filter((session) => !session.isAvailable),
-    ];
```
Comment (ZedLi) [inline ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:26]: Do we still need this logic in the sorted sessions or is it not needed anymore?

**Review by ZedLi — APPROVED** [review]

File: ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js (hunk 5 above)
Comment (lisbet-alvarez) [reply ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:26]: when initially refactoring this, my thoughts were that this logic was unnecessary since users can use the status filter. However, on second thought it doesn't hurt to have additional sorting for when filters are combined.

Comment (ZedLi) [reply ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:26]: I think I agree with you that we probably don't need it but we should just check in with design once they're back. I just wanted to check if this was an explicit decision

Comment (lisbet-alvarez) [reply ui/desktop/app/controllers/scopes/scope/projects/sessions/index.js:26]: I added ICU-12137 to check back in on this.

**Review by cameronperera — APPROVED** [review]
LGTM
